core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1158,8 +1159,8 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
     @throws(classOf[Exception])
     def handleNewSession() {
       info("ZK expired; shut down all controller components and try to re-elect")
+      onControllerResignation()
```

Comment:
Recently I merge this patch into our production env, I found an issue about this:   When the kafka controller startup ( KafkaController.startup) it will lock the controllerlock and subscribe to the zk session expired event.  But if at that time come an zk expired issue (this can happen if kafka startup with a lot of logs reload) before the startup to create the /controller zk node, it will trigger the handleNewSession function,  then it will resign some threads but wait until KafkaController.startup to elect itself as controller and create the /controller node in zk.  After that the onControllerResignation get the controller lock and starting to re-init the  controller context.  And then come the controllerElector.elect, but the /controller node is create the kafka startup not the previous session, so the elect func will skip to elect and start the controller again.  Then the controller thread will gone

@xiguantiaozhan  @guozhangwang  @ijuma